frontend/scripts/react-components/shared/search.component.js

```diff
@@ -0,0 +1,221 @@
+        >
+          {({
+            getInputProps,
+            getItemProps,
+            isOpen,
+            inputValue,
+            highlightedIndex
+          }) => {
+            // stopPropagation is called to avoid calling onOpenClicked.
+            return <div class='autocomplete-container' onClick={e => e.stopPropagation()}>
+              <div class='autocomplete-bar'>
+                <div
+                  style={selectedNodesIds.length > 3 && { overflow: 'auto' }}
+                >
+                  <NodeTitleGroup />
+                </div>
+                <input
+                  {...getInputProps({ placeholder: 'Search a producer, trader or country of import' })}
+                  ref={this.getInputRef}
+                />
+              </div>
+              {isOpen ? (
+                <div class='suggestions'>
+                  {nodes
+                    .filter(
+                      i =>
+                        !inputValue ||
+                        i.name.toLowerCase().includes(inputValue.toLowerCase()),
+                    )
+                    .slice(0, 10)
+                    .map((item, row) => {
+                      // get name segments for highlighting typed string
+                      // ie if you type 'ng', you get ['pi', 'ng', 'po', 'ng']
+                      const nameSegments = findAll({
+                        searchWords: [inputValue],
+                        textToHighlight: item.name
+                      }).map((chunk) => {
+                        const segmentStr = item.name.substr(chunk.start, chunk.end - chunk.start);
+                        return (chunk.highlight) ? <mark>{segmentStr}</mark> : <span>{segmentStr}</span>;
+                      });
+                      const itemSelected = this.isNodeSelected(item);
+                      return (
+                        <div
+                          {...getItemProps({ item })}
+                          key={item.id}
+                          class={cx('suggestion', { '-highlighted': row === highlightedIndex })}
+                        >
+                          <div class='node-text-container'>
+                            <span class='node-type'>{item.type}</span>
+                            <span class='node-name'>
+                              {nameSegments}
+                            </span>
+                          </div>
+                          <div class='node-actions-container'>
+                            <button
+                              onClick={e => this.onAddNode(e, item)}
+                              class='c-button -medium-large'
+                              disabled={itemSelected}
+                            >
+                              {itemSelected ? 'Already in' : 'Add to'} supply chain
+                            </button>
+                            {
+                              PROFILE_PAGES_WHITELIST.includes(item.type) && item.type.split(' & ')
```

Comment:
this is isdented wwwwaaayyyy to deep, which is a smell. I would either:
- split the render method into smaller methods 
- pre-compute all these "map" and "filter" results before. 

We should aim at having JS decoupled from HTML as much as possible, to make things easier to read

Reply:
Agree!